src/js/_enqueues/lib/delayed-inline-script-loader.js

```diff
@@ -0,0 +1,106 @@
+/**
+ * Handles delayed execution of before/after inline scripts for defer/async scripts.
+ *
+ * @output wp-includes/js/wp-delayed-inline-script-loader.js
+ */
+
+(function (window, document) {
+	var nonce = document.currentScript.nonce,
+		doneDependencies = new Set();
+
+	/**
+	 * Determines whether a script was loaded.
+	 *
+	 * @param {string} dep Dependency handle.
+	 * @returns {boolean} Whether dependency was done.
+	 */
+	function isDependencyDone(dep) {
+		return doneDependencies.has(dep);
+	}
+
+	/**
+	 * Runs an inline script.
+	 *
+	 * @param {HTMLScriptElement} script Script to run.
+	 */
+	function runInlineScript(script) {
+		var newScript;
+		script.dataset.wpDone = "1";
+		if (nonce && nonce !== script.nonce) {
+			console.error(
+				"CSP nonce check failed for after inline script. Execution aborted.",
+				script
+			);
+			return;
+		}
+		newScript = script.cloneNode(true);
+		newScript.type = "text/javascript";
+		script.parentNode.replaceChild(newScript, script);
+	}
+
+	/**
+	 * Runs the supplied inline scripts if all of their dependencies have been done.
+	 *
+	 * @param {NodeList<HTMLScriptElement>} scripts Scripts to run if ready.
+	 */
+	function runReadyInlineScripts(scripts) {
+		var i, len, deps;
+		for (i = 0, len = scripts.length; i < len; i++) {
+			deps = scripts[i].dataset.wpDeps.split(/,/);
+			if (deps.every(isDependencyDone)) {
+				runInlineScript(scripts[i]);
+			}
+		}
+	}
+
+	/**
+	 * Runs whenever a load event happens.
+	 *
+	 * @param {Event} event Event.
+	 */
+	function onScriptLoad(event) {
+		var matches, handle, script;
+		if (
+			!(
+				event.target instanceof HTMLScriptElement ||
+				event.target.async ||
+				event.target.defer ||
+				event.target.id
+			)
+		) {
+			return;
+		}
+
+		// TODO: Consider adding a data attribute which specifically contains the handle.
+		matches = event.target.id.match(/^(.+)-js$/);
```

Comment:
You mean another `script` in the page that has that suffix? (Note that the previous condition verifies that the element is a `script`.) But yeah, if that were to happen, it doesn't seem like it would result in any problem since there wouldn't likely be any before/after scripts as well.

If we want to really be sure if this, then we should consider explicitly marking our generated scripts with something like a `data-wp-handle` attribute as mentioned in the comment above. Then we wouldn't have to resort to extracting handles out of the `id`.

Reply:
I do like the ergonomics of not relying on IDs here, but agree that it seems like an extreme edge case that we would have conflicts with those IDs.

Reply:
We can revisit this in #70.